### DISCUSSION TITLE
CRL and OCSP support in TLS

### DISCUSSION THREAD

**Discussion by amdrsantos** [syslog-ng/syslog-ng#4219] [Q&A] [answered]
Is not clear to me if the syslog-ng is supporting the CRL and OCSP validation of the peer's certificate.

For **CRL**, imagine the following scenario while setting up a client / server connection with TLS mutual authentication:
1. Locally no crl-dir is defined
2. Remote certificate has defined CRL in X509 extensions
```
    X509v3 extensions:
         X509v3 CRL Distribution Points: 
                Full Name:
                  URI:http://crl.globalsign.com/gs/gsorganizationvalsha2g2.crl
```
3. Is the syslog-ng downloading the CRL and doing the certificate validation using the information from: http://crl.globalsign.com/gs/gsorganizationvalsha2g2.crl?

A similar scenario for **OCSP**, while setting up a client / server connection with TLS mutual authentication:
If the remote certificate has OCSP server configured in the AIA extension:
```
        X509v3 extensions:
            Authority Information Access: 
                OCSP - URI:http://ocsp2.globalsign.com/gsorganizationvalsha2g2
```
Is the syslog-ng going to send a OCSP request to that server and parse the response?

Other question: Is there a way to add OCSP servers to the TLS configuration?
Is the https://github.com/syslog-ng/syslog-ng/pull/4082 answer to this?

Thanks in advance,
Alex

**Comment by bazsi** [comment, marked as answer]
#4082 is the syslog-ng destination side of OCSP stapling. a syslog-ng destination behaves as an SSL client and validates the server certificate. https://www.ssls.com/knowledgebase/what-is-ocsp-stapling/

We don't perform OCSP requests on our own, in the case of OCSP stapling (e.g. #4082), the server would include the OCSP response and provide it to the client.

Certainly, some functionality is missing, especially when validating the client (e.g. syslog-ng source side)

**Reply by amdrsantos** [reply]
Thanks for the feedback Bazsi.

What about CRL?
Is syslog-ng (client) able to fetch the CRL from a distribution point present in the AIA of the peer's (server) X.509 certificate?

**Reply by bazsi** [reply]
It does not do that either now. It would be possible to do this using an external script, running alongside syslog-ng. Something like this in a shell script or something:
* query the server using openssl s_client, 
* get the server certificate, find out the CRL distribution point, 
* fetch the CRL using wget/curl 
* add it to a directory and set up syslog-ng to use CRLs from that directory.

If this script runs once a day, then CRLs would be kept up-to-date. At the moment, syslog-ng ignores CRL validity related errors  (e.g. if the CRL is not present), but I can easily add that back.

If I wanted to support this with a built-in function, I would rather go to the OCSP stapling direction, e.g. where the server is capable of offering  an OCSP response during the handshake, which the client can verify. The server-side functionality needed for this would be the query the OCSP responder at startup (and then at every expiration) then piggyback this response into the TLS handshake. I am not sure when I would have time for this however and until that happens, the first alternative would be doable without large-scale changes.

**Reply by amdrsantos** [reply]
What is the possibility of having a hook (parameter) to rely on a third party for the certificate validation, like the following example:
```
destination {
    network("test.tld" transport(tls)
        tls(
            pkcs12-file("/path/to/test.p12")
            peer-verify(yes)
            remote-certificate-verification("/usr/bin/cerificate_verification.sh")
            # local-certificate-verification("/usr/bin/cerificate_verification.sh")
        )
    );
};
```
?

**Comment by bazsi** [comment]
This would be a good idea.

**Reply by amdrsantos** [reply]
Is it something that you can do within the next releases?

**Reply by bazsi** [reply]
I hope so. A bit too much on my plate at the moment though.

**Reply by amdrsantos** [reply]
Hi Balázs,

We need this functionality soon, so we try to do the development.
We already have an application to do the OCSP authentication.

Is it possible to implement a solution in a plug-in?
Or it has to be a hook similar to [OCSP stapling](https://github.com/syslog-ng/syslog-ng/pull/4082/files)?

Our solution would be:
1. syslog-ng starts the application and sends the certificate to the stdin of the application.
2. The application does the certificate validation.
3. Now, How can the syslog-ng check if the certificate is valid or not? Using the return code of the application? Could a plug-in be useful here?

How do you see the solution?

Thanks & Regards,
Alex

**Comment by bazsi** [comment]
With stapling the server has to work (and not the client) by acquiring an
ocsp response prior the clients connect and using  that in the server side
of the TLS handshake.

The client side of using such a response is already supported. The server
side is missing.

What you are describing (validating the key on the client against an ocsp
responder) is an alternative to the above. This is doable too and might
support other use-cases, not just ocsp.

With all that said, the first probably needs more C side coding than the
2nd. For the 2nd, you "just" need to fork/exec an external program in the
validation functions (e.g. tls_session_validate_*()).

A somewhat improved version would allow the invocation of plugins on these
cases, so that we could stay in-process if someone wants to. This is
probably more difficult than the first alternative.

My preference would be

1) server side ocsp stapling support
2) proper plugin based hook support on the client side to validate server
certificates
3) validation via fork/exec with an external app

**Reply by amdrsantos** [reply]
Thank you for your support.

For 2)
I understand how to place the hook in lib/transport/tls-session.c, lib/transport/tls-session.h, lib/transport/tls-context.c, lib/transport/tls-context.h

What I am not understanding is how to relate the plugin to the hook.

Can you point me to some examples in the code?

Thanks in Advance,
Alex

**Reply by bazsi** [reply]
Did you get all the help that you need? Do you have questions? Thanks

**Reply by amdrsantos** [reply]
Yes, I have everything for now. Some more priority activities came up. I will be working on the patch/feature afterward.

**Comment by bazsi** [comment]
The main entry point for validating the key for SSL sessions is the
tls_session_verify_callback() function. Apart from doing a number of
validations there, it invokes a "verifier"

```
      if (self->verifier && self->verifier->verify_func)
        return self->verifier->verify_func(ok, ctx,
self->verifier->verify_data);
```

This verifier is set up for socket based destinations in
afinet_dd_setup_tls_verifier() function.

At the moment, this sets up the validation of the hostname. Honestly the
TLSVerifier class could be improved a bit, I find it strange that we need
to pass a function to a constructor, which is then called as a part of a
virtual function. Also, as I see we do create a "VerifyData" object and
then manage its lifetime, instead of simply encapsulating this whole thing
into a class derived from TLSVerifier. But I digress.

So the perfect spot to put the extra validation is in
afinet_dd_verify_callback(). In that function, you get executed once the
server certificate becomes available and it is quite easy to access the
destination driver in which you are running in.

So if you write a function that forks/execs an external program and that
external program can be supplied as an option to the driver, I think we can
address your use-case.

For fork/exec you might be able to something like simply popen()-ing the
app and writing the certificate in DER/PEM format to its stdin. We might as
well send the entire CA chain


On Tue, Jan 3, 2023 at 6:42 PM Alexandre Santos ***@***.***>
wrote:

> Thank you for your support.
>
> For 2)
> I understand how to place the hook in lib/transport/tls-session.c,
> lib/transport/tls-session.h, lib/transport/tls-context.c,
> lib/transport/tls-context.h
>
> What I am not understanding is how to relate the plugin to the hook.
>
> Can you point me to some examples in the code?
>
> Thanks in Advance,
> Alex
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/discussions/4219#discussioncomment-4584245>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5U6TZ4YXTIU5XJMMJLWQRQHVANCNFSM6AAAAAASCTOYV4>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Bazsi

**Reply by amdrsantos** [reply]
If wanted to add a module/plugin, instead of a new function in tls-verifier.c, to do the custom verification, what steps should I make?

**Reply by amdrsantos** [reply]
I already have the hook in place:

```
@@ -143,6 +143,10 @@ afinet_dd_verify_callback(gint ok, X509_STORE_CTX *ctx, gpointer user_data)
       && (tls_context_get_verify_mode(self->tls_context) & TVM_TRUSTED))
     {
       ok = tls_verify_certificate_name(cert, self->hostname);
+      if (ok && self->tls_context->custom_peer_certificate_validation)
+        {
+          ok = tls_verify_certificate_externally(cert, ctx, self->tls_context->custom_peer_certificate_validation);
+        }
     }

     return ok;
```
But instead of calling an executable to lead with the validation, If I want to make a plugin for the validation, can I have the hook in the same place, what else it is needed?

**Comment by Ricfilipe** [comment]
I am trying to achieve a similar thing.
I am trying to extend afscoket module. Is it possible to extend it with a module? If so is there any example of how this can be achieved?

**Reply by MrAnno** [reply]
Additional examples to https://github.com/syslog-ng/syslog-ng/discussions/4219#discussioncomment-5535226 can be found here:

https://github.com/syslog-ng/syslog-ng/tree/master/modules/examples/inner-destinations/http-test-slots
This is an extension module for the HTTP destination.

Here is the HTTP side of it:
https://github.com/syslog-ng/syslog-ng/blob/725cfbfd4a968bb9c2b00fa7f3c6ccd232487dc6/modules/http/http-signals.h
https://github.com/syslog-ng/syslog-ng/blob/725cfbfd4a968bb9c2b00fa7f3c6ccd232487dc6/modules/http/http-worker.c#L228

More info on how to use this example: #3093

----

Or the #4145 pull request also contains some examples of a signal-slot module extension.

**Comment by bazsi** [comment]
If you look at how support for ebpf programs were added, you can notice
that I added a signal to afsocket which is basically a callback function
another plugin can register to.

Note afsocket-signals.h and the Ebpf module that will CONNECT to that
signal. Also there's an EMIT call within afsocket to trigger the hook.

**Comment by Ricfilipe** [comment]
I have created a PR for this #4426
